src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using Microsoft.AspNetCore.Cryptography;
+using Microsoft.AspNetCore.DataProtection.AuthenticatedEncryption;
+using Microsoft.AspNetCore.DataProtection.AuthenticatedEncryption.ConfigurationModel;
+using Microsoft.AspNetCore.DataProtection.SP800_108;
+
+namespace Microsoft.AspNetCore.DataProtection.Managed
+{
+#if NETCOREAPP
+    // An encryptor uses AesGcm to do encryption
+    internal unsafe sealed class AesGcmAuthenticatedEncryptor : IOptimizedAuthenticatedEncryptor, IDisposable
+    {
+        // Having a key modifier ensures with overwhelming probability that no two encryption operations
+        // will ever derive the same (encryption subkey, MAC subkey) pair. This limits an attacker's
+        // ability to mount a key-dependent chosen ciphertext attack. See also the class-level comment
+        // for how this is used to overcome GCM's IV limitations.
+        private const int KEY_MODIFIER_SIZE_IN_BYTES = 128 / 8;
+
+        private const int NONCE_SIZE_IN_BYTES = 96 / 8; // GCM has a fixed 96-bit IV
+        private const int TAG_SIZE_IN_BYTES = 128 / 8; // we're hardcoding a 128-bit authentication tag size
+
+        // 128 "00-01-00-00-00-10-00-00-00-0C-00-00-00-10-00-00-00-10-95-7C-50-FF-69-2E-38-8B-9A-D5-C7-68-9E-4B-9E-2B"
```

Comment:
For these, may help to add a comment referencing the `AesGcmAuthenticatedEncryptor.CreateContextHeader` method. That way future code owners know where the magic headers came from.